fix: resize cursor behavior for frameless windows on Windows

hotdogee opened the pull request:
## Description of Change

**Closes:** #40505
**Related:** microsoft/vscode#185249

## Problem Description

Currently, on Windows, frameless Electron windows (`titleBarStyle: 'hidden'` or `frame: false`) only show the resize cursor and allow resizing when the mouse cursor is exactly on the window border or slightly _inside_ it. This differs from standard framed windows (including Electron's default framed windows) and other applications like Chrome or Edge, where the resize cursor also appears when the cursor is slightly _outside_ the window border. This discrepancy makes resizing frameless windows less intuitive and user-friendly, grabbing the top-right resize handle is especially hard as there is only 1 pixel to grab.

This PR aligns the resizing behavior of frameless windows on Windows with that of standard windows by enabling resizing when the cursor is positioned slightly outside the left, right, and bottom edges.

## Current and Expected Behavior

![frameless-bug-vscode-v2](https://github.com/user-attachments/assets/7799bf27-1805-448d-b3bc-eddac4bc4433)

## Root Cause Analysis

The issue stems from how Windows handles hit-testing for window resizing and how Electron's frameless window implementation interacts with it:

1.  **Windows Hit-Testing (`WM_NCHITTEST`):** Standard window resizing relies on Windows processing the `WM_NCHITTEST` message within the non-client area (the window frame). When the cursor is slightly outside the client area but within the frame's designated resize borders, Windows returns hit-test codes like `HTLEFT`, `HTRIGHT`, `HTBOTTOM`, etc., triggering the resize cursor and behavior.
2.  **Frameless Window Non-Client Area:** For frameless windows, Electron needs to explicitly define the non-client area geometry, typically during the `WM_NCCALCSIZE` message. The previous implementation in `ElectronDesktopWindowTreeHostWin::GetClientAreaInsets` only applied system-defined border thicknesses (`SM_CXSIZEFRAME` + `SM_CXPADDEDBORDER`) when the window was _maximized_. In the _normal_ (non-maximized) state, no such non-client border insets were defined for the left, right, and bottom edges. Consequently, Windows did not receive the necessary geometry information to perform hit-testing just outside the client rectangle for these edges.
3.  **Internal Views Hit-Testing:** Separately, Electron's `FramelessView` performs its own hit-testing (`ResizingBorderHitTestImpl` using `GetHTComponentForFrame`). This logic used a fixed internal inset (`kResizeInsideBoundsSize`) to detect resize handles _inside_ the window bounds.
4.  **Usability issue at the top-right corner:** A significant usability issue arose from the internal hit-testing approach, particularly in the top-right corner. The resize region defined within the window bounds (kResizeInsideBoundsSize) directly overlapped with the caption buttons (e.g., the Close button). In this overlapping zone, the button's hit-test typically took priority, preventing users from initiating a resize from the top-right corner. The previous implementation attempted to mitigate this with a 1-pixel offset for the caption buttons in WinFrameView::LayoutCaptionButtons, essentially carving out a small, dedicated space for resizing. As previously mentioned, this 1 pixel area is extreamly hard to grab for the user.

## Proposed Changes

This pull request modifies the way non-client area calculations and hit-testing are handled for frameless windows on Windows:

1.  **`electron_desktop_window_tree_host_win.cc`:**

    - The primary change is in `GetClientAreaInsets`. Previously, insets (using `GetSystemMetrics(SM_CXSIZEFRAME) + GetSystemMetrics(SM_CXPADDEDBORDER)`) were only applied when the frameless window was maximized.
    - Now, when the window is _not_ maximized, insets are applied to the left, bottom, and right sides (`gfx::Insets::TLBR(0, thickness, thickness, thickness)`). The top inset remains 0 to align with the frameless window behavior.
    - When the window _is_ maximized, the previous behavior of applying insets to all sides is retained (`gfx::Insets::TLBR(thickness, thickness, thickness, thickness)`). This prevents the maximized window from bleeding onto other monitors.
    - These insets are used during the `WM_NCCALCSIZE` message processing. By defining a non-client area border around the client area (even if visually transparent), Windows' default `WM_NCHITTEST` processing can correctly identify `HTLEFT`, `HTRIGHT`, `HTBOTTOM`, etc., when the cursor is in these border regions slightly outside the client rectangle.

2.  **`frameless_view.cc`:**

    - The call to `ResizingBorderHitTestImpl` within `ResizingBorderHitTest` is modified. Instead of passing `gfx::Insets(kResizeInsideBoundsSize)` (which applied the inset uniformly), it now passes `gfx::Insets::TLBR(kResizeInsideBoundsSize, 0, 0, 0)`.
    - This change adjusts the resize detection insets so that the top edge uses the traditional inside-window detection (which works well for the title bar area), while the left, bottom, and right edges get zero inset, allowing the standard Windows behavior of detecting outside the window.

3.  **`win_frame_view.cc`:**
    - Removed the 1px margin adjustment that was previously used as a workaround. With the proper inset handling in the frameless view, this margin is no longer necessary.

## Testing

1.  Built Electron with these code changes on Windows 10, Windows 11, and Ubuntu 24.04.
2.  Integrated the custom Electron build into a VS Code OSS build `1.99.3`.
3.  Tested resizing:
    - Moved the cursor slightly _outside_ the left window edge. **Result:** Resize cursor (↔) appears, and dragging resizes the window horizontally.
    - Moved the cursor slightly _outside_ the right window edge. **Result:** Resize cursor (↔) appears, and dragging resizes the window horizontally.
    - Moved the cursor slightly _outside_ the bottom window edge. **Result:** Resize cursor (↕) appears, and dragging resizes the window vertically.
    - Tested corner resizing (e.g., bottom-right) slightly outside the corner. **Result:** Diagonal resize cursor appears, and dragging resizes correctly.
    - Tested top border resizing. **Result:** Behaves as before (resize cursor appears on/just inside the edge).
    - Maximized the window. **Result:** Resizing is correctly disabled and the maximized window extends to all sides correctly.
    - Restored the window. **Result:** Resizing from outside the borders works again.
    - Tested interaction with Window Controls Overlay (caption buttons): Buttons remain functional and correctly positioned.

https://github.com/user-attachments/assets/6e4ca69e-01e3-4ed2-bd04-66bd5517d9a9

![2025-04-21 193245](https://github.com/user-attachments/assets/fd908aa1-a5e2-4b42-a788-3f47c6fa940d)

## Benefits

- Improved user experience that matches standard Windows application behavior
- Consistent resizing behavior between framed and frameless windows
- No negative impact on other platforms as these changes only affect Windows-specific code

## Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

## Release Notes

Notes: Aligned the resizing behavior of frameless windows with system standards to improve usability on Windows.

codebytere reviewed (approved):
nice work!

hotdogee commented:
Thank you for your feedback! I’ve updated the PR to fix the regression on non‑Windows platforms—my original Ubuntu testing was flawed and has now been corrected. The linting errors have also been resolved.
![image](https://github.com/user-attachments/assets/eff85b65-e52f-4ed1-9acd-873b36163fa5)

codebytere commented:
@hotdogee it looks like some of the failing tests might be relevant - can you validate those locally?

ckerr commented:
The failing tests look relevant to this PR:

```
not ok 194 BrowserWindow module sizing BrowserWindow.setContentSize(width, height) works for a frameless window
  expected [ 440, 559 ] to deeply equal [ 456, 567 ]
  AssertionError: expected [ 440, 559 ] to deeply equal [ 456, 567 ]
      at Context.<anonymous> (electron\spec\api-browser-window-spec.ts:1705:31)
ok 195 BrowserWindow module sizing BrowserWindow.setContentBounds(bounds) sets the content size and position
not ok 196 BrowserWindow module sizing BrowserWindow.setContentBounds(bounds) works for a frameless window
  expected { x: 18, y: 10, width: 234, height: 242 } to deeply equal { x: 10, y: 10, width: 250, height: 250 }
  AssertionError: expected { x: 18, y: 10, width: 234, height: 242 } to deeply equal { x: 10, y: 10, width: 250, height: 250 }
      at expectBoundsEqual (electron\spec\api-browser-window-spec.ts:38:28)
      at Context.<anonymous> (electron\spec\api-browser-window-spec.ts:1731:9)
```

ckerr reviewed (changes requested):
The failing tests need to be resolved, but LGTM in principle

hotdogee commented:
Thanks for the feedback on the failing tests! 🙏
I've updated the pull request to fix failing tests related to `BrowserWindow.setContentSize` and `BrowserWindow.setContentBounds` for frameless windows on Windows.

### Changes Made:
1.  **Size Conversion:** I've updated `NativeWindowViews::ContentBoundsToWindowBounds` and `WindowBoundsToContentBounds`. Now, specifically for frameless, resizable, non-maximized windows on Windows, these functions correctly add/subtract the standard border thickness (`SM_CXSIZEFRAME + SM_CXPADDEDBORDER`) during conversion.
2.  **Resizable Check:** I also added a check in `ElectronDesktopWindowTreeHostWin::GetClientAreaInsets` so that these non-client borders are *only* added if the frameless window is actually `IsResizable()`. Non-resizable windows shouldn't get the resize borders.
3.  **Test Updates:** I've updated the tests in `api-browser-window-spec.ts` that were failing. They now correctly expect that on Windows, the overall window size (`getSize()`) will be larger than the content size (`getContentSize()`) for frameless/hidden title bar windows due to these borders. The expected values reflect the standard border thickness.

These updates resolve the test failures by ensuring accurate bounds calculations on Windows, while maintaining the expected behavior on Linux and macOS.

Please let me know if additional tweaks are needed 🙂

nikwen commented:
Closing this PR because it was abandoned and @mitchchn implemented a fix in https://github.com/electron/electron/pull/50706.
